### PR TITLE
[TE4] Add ping command to python device controller for MRP testing

### LINKED ISSUE #6257: Echo Sever does't work on the applications with chip-shell
### Problem
The application can't correctly respond to the ping command form chip-shell. 
### Description
1.Build the Linux examples in all-clusters-app and lighting-app using
```
gn gen out/debug --args='chip_bypass_rendezvous=true chip_app_use_echo=true'
ninja -C out/debug
```
2.Start the Linux application and send the ping command with chip-shell, the chip-shell receives no response, and the application log is 
```
CHIP:IN: Secure transport received message, but failed to decode it, discarding
```

### PR DESCRIPTION
**Pull request by yufengwangca**
#### Problem
What is being fixed? 
* Currently CSG use chip_shell and chip-echo-responder to test secure channel, and chip-echo-responder has only been enabled on selected platforms with testSecurePairingSecret. we have requirements from CSG to validate secure channel using example apps on real devices over PASE/CASE. We need to find a way to test secure channel and MRP against real example app over PASE/CASE. 
  
* Fixes #6257 Echo Sever does't work on the applications with chip-shell 

#### Change overview
Add ping command to python device controller

#### Testing
How was this tested? (at least one bullet point required)

1. Build chip-lighting-app with echo enabled 
`gn gen out/debug --args='chip_app_use_echo=true'`
2. Start ./chip-lighting-app on the DUT.
3. Run chip-device-ctrl from controller side.
3. Confirm 'ping' command is enabled
```
chip-device-ctrl > ping ?
Usage:

ping [options] <nodeid>

Options:
  -i  <interval>  ping interval time in seconds
  -c  <count>    stop after <count> replies
  -r                   enable MRP
  -s  <size>      payload size in bytes

ping command is using Echo Protocol to measure packet loss across network paths.
```

4. Establish secure connection with device under test
`connect -ip <address> <SetUpPinCode> [<nodeid>]`

5. Ping DUT from python controller
`chip-device-ctrl > ping 12344321`

6. Confirm all echo request have been echoed form DUT
`[1623093477.920869][2573955] CHIP:CTL: Echo Response: 3/3(100.00%) len=32 time=376959.904ms`

### CODE REVIEW

File: src/controller/python/chip-device-ctrl.py
Comment (LuDuda) [inline src/controller/python/chip-device-ctrl.py:563]: Could you update the `README.md` with a new command description under `List of commands` section?

Comment (yufengwangca) [reply src/controller/python/chip-device-ctrl.py:563]: Acked

**Review by yunhanw-google — APPROVED** [review]

File: src/messaging/tests/echo/echo_requester.cpp
Comment (msandstedt) [inline src/messaging/tests/echo/echo_requester.cpp:51]: What's the motivation for this change?

File: src/controller/python/chip-device-ctrl.py
Comment (msandstedt) [inline src/controller/python/chip-device-ctrl.py:579]: I have been told in several PR reviews that refactoring command parsing with e.g. argparse is a great idea, but out-of-scope for a given PR.

But here we have introduction of a brand new command, copying the same manual argument parsing. Why are we doing this?  If we're introducing a brand new command, why can't we use a library parser?

File: src/messaging/tests/echo/echo_requester.cpp
Comment (yufengwangca) [reply src/messaging/tests/echo/echo_requester.cpp]: Good catch, this is from my test change, I should not include it into this PR.

File: src/controller/python/chip-device-ctrl.py
Comment (yufengwangca) [reply src/controller/python/chip-device-ctrl.py:579]: Can argparse be used on parsing the arguments to those sub-functions? I think it is used on parse the arguments to python program.

File: src/controller/CHIPDeviceController.cpp
Comment (tcarmelveilleux) [inline src/controller/CHIPDeviceController.cpp:1040]: Why not use the payload buffer in-place below?

File: src/controller/CHIPDeviceController.cpp
Comment (tcarmelveilleux) [inline src/controller/CHIPDeviceController.cpp:1483]: This does not scale to embedded to use float prints. Is there a way to avoid this showing-up on embedded devices once some have controller ability (e.g. to affect bindings) ?

File: src/controller/CHIPDeviceController.h
Comment (tcarmelveilleux) [inline src/controller/CHIPDeviceController.h:98]: Please use a unit suffix for `mEchoInterval` and `mLastEchoTime`

File: src/controller/CHIPDeviceController.h
Comment (tcarmelveilleux) [inline src/controller/CHIPDeviceController.h:144]:
```suggestion
    // The CHIP Echo interval time in milliseconds.
    uint32_t mEchoIntervalMillis;
```

File: src/controller/CHIPDeviceController.h
Comment (tcarmelveilleux) [inline src/controller/CHIPDeviceController.h:491]:
```suggestion
    CHIP_ERROR PingDevice(NodeId remoteDeviceId, uint32_t maxCount, uint32_t waitTimeMillis, uint32_t payloadSize, bool usingMRP);
```

File: src/controller/python/chip-device-ctrl.py
Comment (tcarmelveilleux) [inline src/controller/python/chip-device-ctrl.py:568]: Why is it seconds here and milliseconds in the API?

File: src/controller/python/chip-device-ctrl.py
Comment (tcarmelveilleux) [reply src/controller/python/chip-device-ctrl.py:579]: Let's use argparse ASAP. It also **properly handles printing default values**.

File: src/controller/python/chip-device-ctrl.py
Comment (tcarmelveilleux) [inline src/controller/python/chip-device-ctrl.py:593]: echoInterval unit needed

File: src/transport/SecureSessionMgr.cpp
Comment (tcarmelveilleux) [inline src/transport/SecureSessionMgr.cpp:201]: This is a very large amount of logging per message. It may impact performance on any embedded class device which has a simplistic logging interface with either blocking entries list, or if it has blocking semantics on full UART buffer or similar. Suggest making the default level of logging here configurable by application

File: src/controller/python/chip-device-ctrl.py
Comment (msandstedt) [reply src/controller/python/chip-device-ctrl.py:579]: This is what I'm suggesting:

```
import argparse
import sys

class ChipTool(object):

    def do_ping(self, line):
        """
        ping [options] <nodeid>
        Options:
          -i  <interval>  ping interval time in seconds
          -c  <count>     stop after <count> replies
          -r  <1|0>       enable or disable MRP
          -s  <size>      payload size in bytes
        ping command is using Echo Protocol to measure packet loss across network paths.
        """
        parser = argparse.ArgumentParser()
        parser.add_argument("-i", type=int, default=1000, dest='interval')
        parser.add_argument("-c", type=int, default=3, dest='count')
        parser.add_argument("-r", type=bool, default=True, dest='mrp')
        parser.add_argument("-s", type=int, default=32, dest='size')
        parser.add_argument('nodeid', type=int, nargs=1)
        args = parser.parse_args(line)
        err = self.devCtrl.PingNode(args['nodeid'], args['count'], args['interval'], args['size'], args['mrp'])

        if err == 0:
            print("Received command status response")
        else:
            print("Failed to receive command status response")

if __name__ == '__main__':

    ChipTool().do_ping(sys.argv[1:])
```

Comment (msandstedt) [reply src/controller/python/chip-device-ctrl.py:579]: > Can argparse be used on parsing the arguments to those sub-functions? I think it is used on parse the arguments to python program.

Yes, as in the above, you can pass an explicit array of arguments to parse_args.  It only defaults to parsing sys.argv if you pass nothing.

File: src/controller/python/chip-device-ctrl.py
Comment (yufengwangca) [reply src/controller/python/chip-device-ctrl.py:568]: This is to align with chip-shell and test spec, we want to use the same test cases when migrating to python device controller.

File: src/controller/CHIPDeviceController.cpp
Comment (tcarmelveilleux) [inline src/controller/CHIPDeviceController.cpp:106]: This is too large. 1280 is the max packet size. The ping payload has to fix after all message layer headers

File: src/controller/CHIPDeviceController.h
Comment (andy31415) [inline src/controller/CHIPDeviceController.h:668]: Is this a spec requirement? 

This feels very out of place,
Why are pingparams static? why dshould every controller support pingparams and echo clients?

File: src/controller/CHIPDeviceController.cpp
Comment (andy31415) [inline src/controller/CHIPDeviceController.cpp:1037]: Use RAII: ScopedBuffer.h

File: src/controller/CHIPDeviceController.cpp
Comment (andy31415) [inline src/controller/CHIPDeviceController.cpp:1077]: you are already returning a CHIP_ERROR, could we log at the caller rather than inline?
IF we do not log here we will not need to use a goto-exit pattern.

File: src/controller/CHIPDeviceController.cpp
Comment (andy31415) [inline src/controller/CHIPDeviceController.cpp:1109]: you can goto-exit instead and place the echo client shutdown in the cleanup block.

Could we RAII the cleanup instead?

File: src/controller/CHIPDeviceController.h
Comment (andy31415) [inline src/controller/CHIPDeviceController.h:132]: please add default initializers for all of these

**Review by andy31415 — CHANGES_REQUESTED** [review]
Marking as request changes for now: this feels very invasive for the scope of "we need testing".
I would like to consider prioritizing code maintainability over having features.

**Comment by yunhanw-google** [comment]
> Marking as request changes for now: this feels very invasive for the scope of "we need testing".
> I would like to consider prioritizing code maintainability over having features.

I think the goal is to start to have ping using different parameter to validate crmp feature and check stability, which could help on code maintability

File: src/controller/CHIPDeviceController.cpp
Comment (yufengwangca) [reply src/controller/CHIPDeviceController.cpp:106]: VerifyOrExit(size <= kMaxPayloadSize, err = CHIP_ERROR_MESSAGE_TOO_LONG);
payloadBuf = MessagePacketBuffer::NewWithData(requestData, size);

This size is used to allocate CHIP message buffer, it contains all message layer headers

**Comment by yufengwangca** [comment]
> > Marking as request changes for now: this feels very invasive for the scope of "we need testing".
> > I would like to consider prioritizing code maintainability over having features.
> 
> I think the goal is to start to have ping using different parameter to validate crmp feature and check stability, which could help on code maintability

This is mandatory requirement from CSG, otherwise, we don't have any way to validate Secure Channel and MRP features on real devices. 

Right now, CSG temporary use echo-responder to test Secure Channel, but echo-responder is only available on Linux and does not use any secure session(PASE/CASE)

File: src/controller/CHIPDeviceController.cpp
Comment (bzbarsky-apple) [reply src/controller/CHIPDeviceController.cpp:106]: `MessagePacketBuffer::NewWithData` will take the passed-in size, add the length of a MIC to that, pass that to PacketBuffer to allocate a buffer with that much payload.  Which will then add the default header size on to that and allocate _that_.

So if `size == kMaxPayloadSize` the above code will try to allocate a PacketBuffer with storage for `1280 + 16 + SystemPacketBuffer::kDefaultHeaderReserve` bytes, right?  Is that expected?

I suspect the right value for `kMAxPayloadSize` is more like `SystemPacketBuffer::kMaxSize - MessagePacketBuffer::kMaxFooterSize` or something.

Or just leave out the size check and just deal with the `NewWithData` failing due to it being too big?

File: src/controller/CHIPDeviceController.cpp
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:1060]:
```suggestion
    ChipLogDetail(Controller, "Send echo request message with payload size: %" PRIu32 " bytes to Node: %" PRIu64, size, remoteDeviceId);
```

Comment (bzbarsky-apple) [reply src/controller/CHIPDeviceController.cpp:1060]: And for the PRIu64, that's not supported on all platforms.  You may want ChipLogFormatX64/ChipLogValueX64

File: src/transport/SecureSessionMgr.cpp
Comment (bzbarsky-apple) [inline src/transport/SecureSessionMgr.cpp:198]: This is logging a struct (which is what `payloadHeader.GetProtocolID()` returns) as %d.  That's not a great idea (and will fail to compile once #7480 lands).  Please fix that and double-check that the other format specifiers here are right.

File: src/controller/CHIPDeviceController.cpp
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:1066]: How do you know you haven't gotten the response yet at this point?  Seems to me we should set this before we make the SendEchoRequest call and unset it on error, or do some sort of synchronization to ensure message processing can't happen under SendEchoRequest.

File: src/controller/python/chip-device-ctrl.py
Comment (yufengwangca) [reply src/controller/python/chip-device-ctrl.py]: Acked

File: src/controller/CHIPDeviceController.cpp
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:1062]: Shouldn't this set to _true_?

File: src/controller/CHIPDeviceController.cpp
Comment (yufengwangca) [reply src/controller/CHIPDeviceController.cpp:1067]: It rely on mPingParams.Reset(); at the beginning of this function to first set this flag to false.  I agree, Reset this flag to false before we make each SendEchoRequest call is better.

File: src/controller/CHIPDeviceController.cpp
Comment (yufengwangca) [reply src/controller/CHIPDeviceController.cpp:1101]: If we go to exist, then we don't need to shutdown the mEchoClient. Yes, we can refactor the EchoClient with RAII, but I would like to leave this refactor in a separate PR.

File: src/controller/CHIPDeviceController.cpp
Comment (bzbarsky-apple) [reply src/controller/CHIPDeviceController.cpp:1067]: No, the point is that you want to set the flag to true _before_ you send the message, as things stand.

File: src/controller/CHIPDeviceController.cpp
Comment (yufengwangca) [reply src/controller/CHIPDeviceController.cpp:1480]: Do we have plan to enable controller on embedded platform?

File: src/controller/CHIPDeviceController.h
Comment (yufengwangca) [reply src/controller/CHIPDeviceController.h:668]: mPingParams is used by OnHandleEchoResponse which need to be static as the callback of EchoClient.

File: src/transport/SecureSessionMgr.cpp
Comment (yufengwangca) [reply src/transport/SecureSessionMgr.cpp]: We also need to update the log on receiving path after #7480 land. Track this with #7503

File: src/controller/CHIPDeviceController.cpp
Comment (yufengwangca) [reply src/controller/CHIPDeviceController.cpp:1062]: why we need to wait for the resp before we send the request?

File: src/controller/CHIPDeviceController.cpp
Comment (yufengwangca) [reply src/controller/CHIPDeviceController.cpp:1067]: I am confused, the current logic is to reset the WaitingForEchoResp flag to false before we send the echo request, and change it to true after the echo request has been successfully sent.

File: src/controller/CHIPDeviceController.cpp
Comment (yufengwangca) [reply src/controller/CHIPDeviceController.cpp]: Acked

File: src/controller/CHIPDeviceController.cpp
Comment (yufengwangca) [reply src/controller/CHIPDeviceController.cpp]: I think we can just use buffer on stack for this small message content, no need to use dynamic memory.

File: src/controller/CHIPDeviceController.cpp
Comment (yufengwangca) [reply src/controller/CHIPDeviceController.cpp]: I prefer leave it to the size check within NewWithData

File: src/controller/CHIPDeviceController.cpp
Comment (bzbarsky-apple) [reply src/controller/CHIPDeviceController.cpp:1062]: Because in the current state of the tree the response might come back before the send call returns, and then this logic will be broken.  So as a workaround we can set the waiting flag to true first, then set it to false if send fails, which will be more robust to the ordering stuff...

File: src/controller/CHIPDeviceController.cpp
Comment (bzbarsky-apple) [reply src/controller/CHIPDeviceController.cpp:1067]: See:

> How do you know you haven't gotten the response yet at this point?

And the answer, given our current threading situation is that you don't.

File: src/controller/CHIPDeviceController.cpp
Comment (yufengwangca) [reply src/controller/CHIPDeviceController.cpp]: Now, I got your point

File: src/controller/CHIPDeviceController.cpp
Comment (yufengwangca) [reply src/controller/CHIPDeviceController.cpp]: Acked

File: src/controller/CHIPDeviceController.cpp
Comment (andy31415) [reply src/controller/CHIPDeviceController.cpp:1104]: I would understand changing existing code to be out of scope for refactors - don't want to both add functionality and refactor in the same PR.

However this is new code, so I expect it to be withouth gotoexit to start with.

File: src/controller/CHIPDeviceController.h
Comment (andy31415) [reply src/controller/CHIPDeviceController.h:668]: Please split out echo handling more - I believe "commisioner" should not be directly responsible for echo/ping parameters and callbacks. If some integration is required, please make it minimal. OnHandle.... should not be part of the commisioner.

File: src/controller/CHIPDeviceController.cpp
Comment (yufengwangca) [reply src/controller/CHIPDeviceController.cpp]: Acked

File: src/controller/CHIPDeviceController.h
Comment (yufengwangca) [reply src/controller/CHIPDeviceController.h]: Acked, refactored the Callback handler in Echo with delegate pattern, we don't need to use the static member functions and variables for handling the callback.

File: src/controller/python/chip-device-ctrl.py
Comment (msandstedt) [reply src/controller/python/chip-device-ctrl.py]: Thanks for the refactor.  Hopefully others can replicate the pattern when adding new commands.

File: src/controller/CHIPDeviceController.cpp
Comment (msandstedt) [reply src/controller/CHIPDeviceController.cpp]: A lot of the patterns in the controller object (e.g. statically allocated memory for controlled devices) imply we are trying to accommodate embedded platform controllers.  RTOS-based Thread border routers seem a likely candidate.  More generally, the device-to-device model would require this be supportable.

File: src/controller/PingClient.cpp
Comment (msandstedt) [inline src/controller/PingClient.cpp:124]: Thanks for including this comment.

It does feel odd to have blocking code in `src/controller`.  I would be less concerned about such a thing in `./examples`.

**Review by msandstedt — APPROVED** [review]

File: src/controller/CHIPDeviceController.cpp
Comment (yufengwangca) [reply src/controller/CHIPDeviceController.cpp]: acked

File: src/controller/PingClient.cpp
Comment (yufengwangca) [reply src/controller/PingClient.cpp]: Acked, made ping device operation asynchronously

File: src/controller/PingClient.h
Comment (cecille) [inline src/controller/PingClient.h:33]: Does this handle both IPv4 and IPv6 ping? Could you add a comment to clarify where this is going out? Is there an address preference?

**Comment by yunhanw-google** [comment]
please also add test support in ./src/controller/python/test/test_scripts/mobile-device-test.py

**Comment by yufengwangca** [comment]
> please also add test support in ./src/controller/python/test/test_scripts/mobile-device-test.py

Prefer to update cirque script with this command in a separate PR

File: src/controller/PingClient.h
Comment (yufengwangca) [reply src/controller/PingClient.h:33]: This ping is on the application level, not on the ip level. It is depending on how to connect the peer device before you issue the ping command.

File: src/controller/PingClient.cpp
Comment (andy31415) [inline src/controller/PingClient.cpp:21]: If the protocol is "Echo" why is this called PingClient? Why not EchoClient or PacketLossMeasureClient?

File: src/controller/CHIPDeviceController.h
Comment (andy31415) [inline src/controller/CHIPDeviceController.h:621]: I understand DeviceCommisioner as a "extension to DeviceController that allows commisioning functions, i.e. allows for PASE establishment".

By that definition, PingClient has no place here. 
I was trying to read through the spec and generally could not find "echo" or "ping" so have a hard time placing it. If this is specifically for test events to validate the example apps, this should be part of examples. However if the spec does not mandate it, I would like to understand why this is mandatory for tests and how does it fit in - if this is not a spec requirement, why is it part of the certification?

File: src/controller/PingClient.h
Comment (andy31415) [inline src/controller/PingClient.h:82]: initialize to nullptr, this and below.

File: src/controller/PingClient.h
Comment (cecille) [reply src/controller/PingClient.h:33]: If this isn't actually a ping, could it be called something else? Ping is pretty specific.

**Comment by yufengwangca** [comment]
Need to look for alternative spec compliant solution for MRP/BTP validation based on discussion offline

**Pull request closed without merging**
